### ISSUE TITLE
Handle Multiple Authentication Types

### ISSUE THREAD

**Issue by logikone** [clastix/capsule-proxy#258]
We currently are attempting to expose capsule-proxy via [Contour](https://projectcontour.io). Contour is configured to send a client cert with each request proxied to a backend service to allow the backend service to validate that the request is coming from an expected source, which capsule-proxy has support for. However the cert is supplied by capsule and not by the client so there are no username/group claims present on the cert.

I'd like to be able to somehow specify which authentication type we are wanting to use, or be able to configure the order in which types are tried [here](https://github.com/clastix/capsule-proxy/blob/v0.3.2/internal/request/http.go#L138-L147). 

I can do the work just looking for some feedback on what the best way to accomplish this would be.

Thanks!

**Comment by sagar-jadhav** [comment]
How about addressing this use case as follows:

- Introducing a new authType `certificateAndBearerBased` [here](https://github.com/clastix/capsule-proxy/blob/b8c26d7f1882463fbc702dc1b2a66cb5f29cd4bf/internal/request/http.go#L138-L147)
- This authType is return when request contains both TLS data and bearer token
- In [GetUserAndGroups](https://github.com/clastix/capsule-proxy/blob/b8c26d7f1882463fbc702dc1b2a66cb5f29cd4bf/internal/request/http.go#L41) func introduce one more switch case for `certificateAndBearerBased` authType. In this case first try retrieving username and group info from TLS cert If that info is not available in the TLS cert then use bearer token to get that information.

Basically with this approach we are introducing a new auth flow rather than touching the existing flow.

**Comment by rakshpatel** [comment]
Can we not extend the check in **getAuthType()** function for same purpose stated above instead of adding a new **authType**?

**Comment by prometherion** [comment]
Deciding the order of validation is always really opinionated, couldn't we go Kubernetes-ish here, as `kube-apiserver` is doing with `--kubelet-preferred-address-types`?

Just a proposal, `--auth-preferred-types` with comma-separated enums, such as `TLSCertificate`, `BearerToken` not sure if there are others, TBH, although I remember some issues with `pinniped` @slushysnowman?

To avoid breaking change compatibility issues, the default value would be `TLSCertificate,BearerToken`.

Strong opinions, weakly held: let me know if this idea is reasonable or not without considering I'm one of the maintainers 🙃

**Comment by sagar-jadhav** [comment]
Setting` --auth-preferred-types` makes sense here. As cluster admin you are aware about the infrastructure and supported auth types so accordingly you can override the value of `--auth-preferred-types` (If needed).

**Comment by rakshpatel** [comment]
Well, yes that makes more sense and will be flexible without much code change i'm guessing.

**Comment by prometherion** [comment]
`pflags` doesn't support enum 🥲 

But it should be quite easy, famous last words.

@logikone please, share your feedback on this proposal too, your feedback is absolutely valuable, and thanks for giving a chance to Capsule and its proxy! 🥳

**Comment by logikone** [comment]
I've used [thediveo/enumflag](https://github.com/thediveo/enumflag) for enums with pflag before and that worked fine. I can add constants and an enum flag with the default set to `TLSCertificate,BearerToken`

**Comment by prometherion** [comment]
If you're open to contribute, please be my guest, it would be great! Assigning the issue to you, let me know if you have any issue.

**Comment by logikone** [comment]
awesome, ya i'll get going on a pr. ty sir!
